### PR TITLE
DCOS-10984: volume mounts

### PR DESCRIPTION
**Pull request by Poltergeist**
~~based on https://github.com/dcos/dcos-ui/pull/1528~~ Merged

### CODE REVIEW

File: plugins/services/src/js/reducers/serviceForm/VolumeMounts.js [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+        break;
+    }
+
+    if (path[2] === 'name') {
+      newState[path[1]].name = value;
+    }
+
+    return newState;
+  },
+  JSONParser(state) {
```
Comment (nLight) [inline plugins/services/src/js/reducers/serviceForm/VolumeMounts.js]: There's some magic happening in this function let's test it

File: plugins/services/src/js/reducers/serviceForm/VolumeMounts.js [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+          containerIndex
+        ], mount.mountPath));
+
+        return memo;
+      }, []));
+
+      return memo;
+    }, []));
+
+    return memo;
```
Comment (nLight) [inline plugins/services/src/js/reducers/serviceForm/VolumeMounts.js]: I bet you've heard I like `reduce`, so you nested 3 of them...
Would be awesome to have some comments in the code

File: plugins/services/src/js/components/modals/NewCreateServiceModal.js [hunk 3]
```diff
@@ -325,7 +329,11 @@ class NewServiceFormModal extends Component {
         {
           cmd: !isPod,
           container: !isPod,
-          containers: isPod
+          containers: isPod,
```
Comment (nLight) [inline plugins/services/src/js/components/modals/NewCreateServiceModal.js]: can we split this into 2 lists: `PodReducers` `ServiceReducers`? Because now containers is a pod reducer surrounded by non-pod reducers and I didn't catch that right away.

File: plugins/services/src/js/reducers/serviceForm/MultiContainerNetwork.js [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+      }
+
+      return {
+        mode: valueSplit[0].toLowerCase()
+      };
+    }
+    return state;
+  },
+  JSONParser(state) {
+    if (state == null || state.network == null || state.network.mode == null) {
```
Comment (nLight) [inline plugins/services/src/js/reducers/serviceForm/MultiContainerNetwork.js]: I'd rather use `Util.findNestedSomething..`

File: plugins/services/src/js/reducers/serviceForm/VolumeMounts.js [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+
+    return newState;
+  },
+  JSONParser(state) {
+    if (state == null) {
+      return [];
+    }
+
+    let volumes = [];
+    let mountMap = {};
```
Comment (nLight) [inline plugins/services/src/js/reducers/serviceForm/VolumeMounts.js]: rename suggestion `volumesNameToIndex`?

**Comment by nLight** [comment]
This is a great piece of work and it is actually working (I tested), please consider my comments and 🚢 !

File: plugins/services/src/js/reducers/serviceForm/VolumeMounts.js (hunk 2 above)
Comment (mlunoe) [inline plugins/services/src/js/reducers/serviceForm/VolumeMounts.js]: Ewww... could we make that a little simpler, please

File: plugins/services/src/js/components/modals/NewCreateServiceModalForm.js [hunk 6]
```diff
@@ -338,8 +342,15 @@ class NewCreateServiceModalForm extends Component {
               onRemoveItem={this.handleRemoveItem}
               onAddItem={this.handleAddItem} />
         </TabView>,
-        <TabView id="networking" key="networking">
-          <NetworkingFormSection
+        <TabView id="volumes" key="multivolumes">
+          <VolumeMountFormSection
+              data={data}
```
Comment (mlunoe) [inline plugins/services/src/js/components/modals/NewCreateServiceModalForm.js]: Indented one too much. I've seen this _a lot_ lately. Could you fix the editor config or your editor settings?

**Comment by jfurrow** [comment]
I noticed a few problems:
1. Adding a volume and then editing it in the JSON editor seems to duplicate the volume in the form.


2. Adding a volume with the `mountPath` set to a value, and then removing the volume, leaves the values in the `containers.volumeMounts` array.

I pushed a minor change regarding the use of `FormGroupContainer` :)

File: plugins/services/src/js/components/forms/MultiContainerNetworkingFormSection.js [hunk 7]
```diff
@@ -0,0 +1,451 @@
...
+
+  getVirtualNetworks() {
+    return VirtualNetworksStore.getOverlays().mapItems((overlay) => {
+      let name = overlay.getName();
+
+      return {
+        text: `Virtual Network: ${name}`,
+        value: `${Networking.type.CONTAINER}.${name}`
+      };
+    }).getItems().map((virtualNetwork, index) => {
```
Comment (mlunoe) [inline plugins/services/src/js/components/forms/MultiContainerNetworkingFormSection.js]: Can't we just do `VirtualNetworksStore.getOverlays().getItems().map(...)`?

File: plugins/services/src/js/components/forms/MultiContainerNetworkingFormSection.js [hunk 8]
```diff
@@ -0,0 +1,451 @@
...
+      `containers.${containerIndex}.endpoints.${index}.port`
+    );
+
+    if (endpoint.automaticPort) {
+      placeholder = `$PORT${index}`;
+    }
+
+    let tooltipContent = (
+      <span>
+        {'This host port will be accessible as an environment variable called `$PORT{index}0`. '}
```
Comment (mlunoe) [inline plugins/services/src/js/components/forms/MultiContainerNetworkingFormSection.js]: This string is not evaluated and should probably be:
```js
{`This host port will be accessible as an environment variable called $PORT${index}. `}
```

File: plugins/services/src/js/components/forms/MultiContainerNetworkingFormSection.js [hunk 9]
```diff
@@ -0,0 +1,451 @@
...
+import HostUtil from '../../utils/HostUtil';
+import Icon from '../../../../../../src/js/components/Icon';
+import Networking from '../../../../../../src/js/constants/Networking';
+import VirtualNetworksStore from '../../../../../../src/js/stores/VirtualNetworksStore';
+
+const METHODS_TO_BIND = [
+  'onVirtualNetworksStoreSuccess'
+];
+
+class MultiContainerHealthChecksFormSection extends mixin(StoreMixin) {
```
Comment (mlunoe) [inline plugins/services/src/js/components/forms/MultiContainerNetworkingFormSection.js]: There is a lot of this section that is just a copy of `NetworkingFormSection`. I thought we aimed to build section parts that we could re-use, so we would not end up in this situation. E.g fixing here, without knowing that the same issue is in `NetworkingFormSection`, which then will be forgotten.

File: plugins/services/src/js/components/forms/VolumeMountFormSection.js [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+import FieldError from '../../../../../../src/js/components/form/FieldError';
+import FieldInput from '../../../../../../src/js/components/form/FieldInput';
+import FieldLabel from '../../../../../../src/js/components/form/FieldLabel';
+import FormGroup from '../../../../../../src/js/components/form/FormGroup';
+import FormGroupContainer from '../../../../../../src/js/components/form/FormGroupContainer';
+import {FormReducer as volumeMounts} from '../../reducers/serviceForm/VolumeMounts';
+
+const errorsLens = Objektiv.attr('container', {}).attr('volumes', []);
+
+class VolumesFormSection extends Component {
```
Comment (mlunoe) [inline plugins/services/src/js/components/forms/VolumeMountFormSection.js]: This again seems very similar to, but much more advanced than `VolumesFormSection` (it's even still named the same).
Could we not avoid this duplication, and create small section parts that they both can use?

File: plugins/services/src/js/reducers/serviceForm/MultiContainerNetwork.js [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+      const mode = Networking.type.CONTAINER;
+      const name = findNestedPropertyInObject(state, 'network.name');
+
+      return new Transaction(['network'], `${mode}.${name}`);
+    }
+
+    return new Transaction(['network'], state.network.mode.toUpperCase());
+  },
+  FormReducer(state = {mode: Networking.type.HOST}, {type, path, value}) {
+    const joinedPath = path.join('.');
```
Comment (mlunoe) [inline plugins/services/src/js/reducers/serviceForm/MultiContainerNetwork.js]: Remember to either default assign `path` or do:
```js
if (path == null) {
  return state;
}
```

File: plugins/services/src/js/reducers/serviceForm/MultiContainerNetwork.js [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+
+    if (joinedPath === 'network' && type === SET) {
+      const valueSplit = value.split('.');
```
Comment (mlunoe) [inline plugins/services/src/js/reducers/serviceForm/MultiContainerNetwork.js]: `const [mode, name] = value.split('.');`?

File: plugins/services/src/js/reducers/serviceForm/MultiContainerNetwork.js [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+
+      return new Transaction(['network'], `${mode}.${name}`);
+    }
+
+    return new Transaction(['network'], state.network.mode.toUpperCase());
+  },
+  FormReducer(state = {mode: Networking.type.HOST}, {type, path, value}) {
+    const joinedPath = path.join('.');
+
+    if (joinedPath === 'network' && type === SET) {
```
Comment (mlunoe) [inline plugins/services/src/js/reducers/serviceForm/MultiContainerNetwork.js]: Maybe check that value is actually an array: `&& Array.isArray(value)`

File: plugins/services/src/js/reducers/serviceForm/VolumeMounts.js [hunk 14]
```diff
@@ -0,0 +1,133 @@
+import {ADD_ITEM, REMOVE_ITEM} from '../../../../../../src/js/constants/TransactionTypes';
+import Transaction from '../../../../../../src/js/structs/Transaction';
+
+module.exports = {
+  JSONReducer(state = [], {type, path, value}) {
+    if (path[0] !== 'volumeMounts') {
```
Comment (mlunoe) [inline plugins/services/src/js/reducers/serviceForm/VolumeMounts.js]: `const [base, index, name] = path`?

File: plugins/services/src/js/reducers/serviceForm/VolumeMounts.js [hunk 15]
```diff
@@ -0,0 +1,133 @@
...
+  JSONParser(state) {
+    if (state == null) {
+      return [];
+    }
+
+    let volumes = [];
+    let mountMap = {};
+
+    let memo = [];
+    if (state.volumes != null) {
```
Comment (mlunoe) [inline plugins/services/src/js/reducers/serviceForm/VolumeMounts.js]: `Array.isArray(state.volumes)`, please

File: plugins/services/src/js/reducers/serviceForm/VolumeMounts.js [hunk 16]
```diff
@@ -0,0 +1,133 @@
...
+  },
+  JSONParser(state) {
+    if (state == null) {
+      return [];
+    }
+
+    let volumes = [];
+    let mountMap = {};
+
+    let memo = [];
```
Comment (mlunoe) [inline plugins/services/src/js/reducers/serviceForm/VolumeMounts.js]: Could we rename the outer `memo`. E.g. `volumeTransactions`?

File: plugins/services/src/js/reducers/serviceForm/VolumeMounts.js [hunk 17]
```diff
@@ -0,0 +1,133 @@
...
+    if (state.volumes != null) {
+      memo.push(...state.volumes.reduce((memo, volume) => {
```
Comment (mlunoe) [inline plugins/services/src/js/reducers/serviceForm/VolumeMounts.js]: Let's split this into readable statements:
```js
const mountTransactions = state.volumes.reduce((memo, volume) => {
  ...
});
volumeTransactions = volumeTransactions.concat(mountTransactions);
```

File: plugins/services/src/js/reducers/serviceForm/VolumeMounts.js [hunk 18]
```diff
@@ -0,0 +1,133 @@
...
+        return memo;
+      }, []));
+    }
+
+    memo.push(...state.containers.reduce((memo, container, containerIndex) => {
+      if (container.volumeMounts == null) {
+        return memo;
+      }
+
+      memo.push(...container.volumeMounts.reduce((memo, mount) => {
```
Comment (mlunoe) [inline plugins/services/src/js/reducers/serviceForm/VolumeMounts.js]: let's use concat for these

File: plugins/services/src/js/reducers/serviceForm/VolumeMounts.js (hunk 2 above)
Comment (mlunoe) [inline plugins/services/src/js/reducers/serviceForm/VolumeMounts.js]: I mean; split it into separate functions if you have to. This is impossible to follow, what is going on

File: plugins/services/src/js/reducers/serviceForm/VolumeMounts.js [hunk 19]
```diff
@@ -0,0 +1,133 @@
...
+      }, []));
+
+      return memo;
+    }, []));
+
+    return memo;
+  },
+  FormReducer(state = [], {type, path, value}) {
+    let newState = state.slice();
+
```
Comment (mlunoe) [inline plugins/services/src/js/reducers/serviceForm/VolumeMounts.js]: `const [base, index, name] = path`?

File: plugins/services/src/js/components/forms/MultiContainerEnvironmentFormSection.js [hunk 20]
```diff
@@ -0,0 +1,162 @@
...
+
+import {FormReducer as env} from '../../reducers/serviceForm/EnvironmentVariables';
+import {FormReducer as labels} from '../../reducers/serviceForm/Labels';
+import DeleteRowButton from '../../../../../../src/js/components/form/DeleteRowButton';
+import FieldError from '../../../../../../src/js/components/form/FieldError';
+import FieldInput from '../../../../../../src/js/components/form/FieldInput';
+import FieldLabel from '../../../../../../src/js/components/form/FieldLabel';
+import FormGroup from '../../../../../../src/js/components/form/FormGroup';
+
+class MultiContainerEnvironmentFormSection extends Component {
```
Comment (mlunoe) [inline plugins/services/src/js/components/forms/MultiContainerEnvironmentFormSection.js]: The only difference I see to this and `EnvironmentFormSection` is the name and the Mount.
Let re-use please

**Comment by orlandohohmeier** [comment]
Regarding the duplication issue that @jfurrow mentioned I'd say that even though that it seems wrong at the first glance it looks like the least error prone attempt to deal with different volume/mount values.

**Comment by orlandohohmeier** [comment]
We need to add a few info blocks to the form as the volume section looks highly confusing at the moment...

**Comment by nLight** [comment]
rebased on https://github.com/dcos/dcos-ui/pull/1528

File: plugins/services/src/js/components/forms/MultiContainerNetworkingFormSection.js (hunk 8 above)
Comment (nLight) [inline plugins/services/src/js/components/forms/MultiContainerNetworkingFormSection.js]: fixed in https://github.com/dcos/dcos-ui/pull/1528

File: plugins/services/src/js/reducers/serviceForm/MultiContainerNetwork.js (hunk 12 above)
Comment (nLight) [inline plugins/services/src/js/reducers/serviceForm/MultiContainerNetwork.js]: fixed https://github.com/dcos/dcos-ui/pull/1528

File: plugins/services/src/js/components/forms/VolumeMountFormSection.js (hunk 10 above)
Comment (nLight) [inline plugins/services/src/js/components/forms/VolumeMountFormSection.js]: We keep duplication for now

File: plugins/services/src/js/components/forms/MultiContainerNetworkingFormSection.js (hunk 7 above)
Comment (nLight) [inline plugins/services/src/js/components/forms/MultiContainerNetworkingFormSection.js]: Let's postpone this

File: plugins/services/src/js/components/modals/NewCreateServiceModalForm.js (hunk 6 above)
Comment (nLight) [inline plugins/services/src/js/components/modals/NewCreateServiceModalForm.js]: Let's improve our eslint rules 💪

File: plugins/services/src/js/reducers/serviceForm/VolumeMounts.js (hunk 19 above)
Comment (nLight) [inline plugins/services/src/js/reducers/serviceForm/VolumeMounts.js]: done

File: plugins/services/src/js/reducers/serviceForm/VolumeMounts.js (hunk 18 above)
Comment (nLight) [inline plugins/services/src/js/reducers/serviceForm/VolumeMounts.js]: done

File: plugins/services/src/js/components/modals/NewCreateServiceModalForm.js (hunk 6 above)
Comment (orlandohohmeier) [inline plugins/services/src/js/components/modals/NewCreateServiceModalForm.js]: ➡️  #1565

**Comment by rcorral** [comment]
Could you move `Volumes` right under the networking tab `Networking` like on the designs
img

**Comment by rcorral** [comment]
Could you also match the form styles like the designs please
img

File: plugins/services/src/js/reducers/serviceForm/VolumeMounts.js (hunk 5 above)
Comment (mlunoe) [inline plugins/services/src/js/reducers/serviceForm/VolumeMounts.js]: Went with a compromise: `volumeIndexMap`

**Comment by mlunoe** [comment]
Fixed here:

**Comment by mlunoe** [comment]
Retest this please

File: plugins/services/src/js/reducers/serviceForm/Containers.js [hunk 21]
```diff
@@ -323,11 +323,13 @@ module.exports = {
...
-    this.volumeMounts =
-        volumeMountsReducer(this.volumeMounts, {type, path, value});
+    this.volumeMounts = volumeMountsReducer(
+      this.volumeMounts,
+      {type, path, value}
+    );
 
     newState = state.map((container, index) => {
-      if (this.volumeMounts.length !== 0) {
+      if (this.volumeMounts.length === 0) {
```
Comment (nLight) [inline plugins/services/src/js/reducers/serviceForm/Containers.js]: I'm not sure why did you change the condition? When `volumeMounts.length === 0` we don't have anything to filter nor map. Do we?

File: plugins/services/src/js/reducers/serviceForm/Containers.js [hunk 22]
```diff
@@ -438,7 +443,8 @@ module.exports = {
   },
 
   FormReducer(state, {type, path = [], value}) {
-    const [base, index, field, secondIndex, name] = path; // eslint-disable-line no-unused-vars
+    // eslint-disable-next-line no-unused-vars
+    const [_, index, field, secondIndex, name] = path;
```
Comment (nLight) [inline plugins/services/src/js/reducers/serviceForm/Containers.js]: 👍 thanks!

Comment (nLight) [inline plugins/services/src/js/reducers/serviceForm/Containers.js]: in general I prefer having underscore prefix ignored in eslint so it would be
`const [_base, index, field, secondIndex, name] = path;` Shall we bring this to the next sync?

File: plugins/services/src/js/reducers/serviceForm/VolumeMounts.js [hunk 23]
```diff
@@ -10,8 +10,9 @@ function transformContainers(memo, container, containerIndex) {
     return memo;
   }
 
-  const tuples =
-    container.volumeMounts.map((mount) => ([containerIndex, mount]));
+  const tuples = container.volumeMounts.map((mount) => {
```
Comment (nLight) [inline plugins/services/src/js/reducers/serviceForm/VolumeMounts.js]: you are right, tendency to keep everything in one line is not good here

File: plugins/services/src/js/reducers/serviceForm/VolumeMounts.js [hunk 24]
```diff
@@ -57,45 +58,40 @@ module.exports = {
...
-        memo.push(new Transaction(['volumeMounts'], mountMap[volume.name], ADD_ITEM));
-        memo.push(new Transaction([
-          'volumeMounts',
-          mountMap[volume.name],
-          'name'
-        ], volume.name));
-
-        return memo;
+        volumeIndexMap[volume.name] = volumes.push(volume.name) - 1;
+        return memo.concat(
```
Comment (nLight) [inline plugins/services/src/js/reducers/serviceForm/VolumeMounts.js]: 👍

File: plugins/services/src/js/reducers/serviceForm/VolumeMounts.js [hunk 25]
```diff
@@ -57,45 +58,40 @@ module.exports = {
...
-
-          memo.push(new Transaction([
-            'volumeMounts',
-            mountMap[name],
-            'mountPath',
-            containerIndex
-          ], mountPath));
+    const containers = state.containers || [];
+    const containerVolumesTransactions = containers
+      .reduce(transformContainers, [])
```
Comment (nLight) [inline plugins/services/src/js/reducers/serviceForm/VolumeMounts.js]: How did you like the approach in general? Still not sure

**Comment by nLight** [comment]
I fixed the bug, LGTM 🚀

File: plugins/services/src/js/reducers/serviceForm/VolumeMounts.js (hunk 25 above)
Comment (mlunoe) [inline plugins/services/src/js/reducers/serviceForm/VolumeMounts.js]: Yeah, the nested reduce is not pretty. I feel we can break this into separate functions, but I think we have a few functions like this, that needs a cleanup. I posted a few comments about it here too: https://github.com/dcos/dcos-ui/pull/1524#pullrequestreview-12070726

File: plugins/services/src/js/reducers/serviceForm/Containers.js (hunk 21 above)
Comment (mlunoe) [inline plugins/services/src/js/reducers/serviceForm/Containers.js]: Sorry, this was a mistake. Thanks for the catch!
